boards: intel_s1000: Image download scripts

sathishkuttan opened the pull request:
This PR adds python based scripts to be used on a linux development host that is connected over SPI to Intel S1000 board.
The script accepts a zephyr.bin image and downloads it to the target. The target then executes the image.

Commit:
boards: intel_s1000_crb: Host interface config

A YAML file to specify configuration of interface I/Os
such as the SPI device, GPIOs, etc.
The image download script configures the host's SPI and GPIO
interfaces accordingly.

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

Commit:
boards: intel_s1000_crb: Image download scripts

Python scripts to download a zephyr binary image (zephyr.bin) to
Intel S1000 from a linux host.
The linux host's SPI master and GPIOs shall be connected to the
corresponding SPI slave and I/Os respectively.

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

nashif commented:
@sathishkuttan can you add a section in the board documentation describing how this can be used?

Commit:
doc: intel_s1000: image download from Linux host

Documentation update to add instructions to download a
zephyr binary to the target from a Linux host

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

sathishkuttan commented:
>@sathishkuttan can you add a section in the board documentation describing how this can be used?

@nashif done

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/index.rst:
```diff
@@ -173,6 +173,25 @@ application.
+
+.. code-block:: console
+
+   sudo pip3 install pyyaml python-periphery hashlib bitstruct
```
We've been recommending folks use ``--user`` rather than using ``sudo`` to install stuff with pip:
```
pip3 install --user pyyaml python-periphery hashlib bitstruct
```

sathishkuttan replied:
> We've been recommending folks use `--user` rather than using `sudo` to install stuff with pip:
> 
> ```
> pip3 install --user pyyaml python-periphery hashlib bitstruct
> ```

Changed.

dcpleung replied:
The download.py would probably not work at all since it is run as root, so these packages would not be visible to the script.

sathishkuttan replied:
> The download.py would probably not work at all since it is run as root, so these packages would not be visible to the script.

suggested solution? go back to installing the packages via `sudo pip3`?

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/index.rst:
```diff
@@ -173,6 +173,26 @@ application.
+
+.. code-block:: console
+
+   pip3 install --user pyyaml python-periphery hashlib bitstruct
```
After the code block add:
```
Verify that ``~/.local/bin`` is at the beginning of your PATH (and add it
if it's missing) so these python modules can be found.
```

sathishkuttan replied:
Aren't these generic instructions for installation of python modules? It doesnt make sense to duplicate these in every doc page. If there is page that already has these instructions, I would rather point to it.

dbkinder replied:
Problem is that folks aren't reading the documentation when a link is provided. As a usability improvement, it's good to put critical information near where it's needed.  (I'm hunting down other stray uses of ``pip3 install --user`` and updating those too.)

dbkinder replied:
On second thought, you can leave this as is.  I'll come back around if needed to hunt down these uses of --user.

dbkinder commented on boards/xtensa/intel_s1000_crb/doc/index.rst:
```diff
@@ -173,6 +173,26 @@ application.
+.. code-block:: console
+
+   cd <app-dir>/build
+   sudo -E python3 \
```
it reads a bit better if the continuation lines are indented:
```
   sudo -E python3 \
      $ZEPHYR_BASE/boards/xtensa/intel_s1000_crb/support/download.py \
      zephyr/zephyr.bin
```

dbkinder commented on boards/xtensa/intel_s1000_crb/support/device.py:
```diff
@@ -0,0 +1,82 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Author: Sathish Kuttan <sathish.k.kuttan@intel.com>
+
```
Would be good to include some documentation comments here about what this script is for (same goes for the other Python scripts in this PR).

dbkinder commented on boards/xtensa/intel_s1000_crb/support/download.py:
```diff
@@ -0,0 +1,103 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# Author: Sathish Kuttan <sathish.k.kuttan@intel.com>
+
```
Comments explaining what this Python script does would be a good addition.

dbkinder commented on boards/xtensa/intel_s1000_crb/support/download.py:
```diff
@@ -0,0 +1,103 @@
+
+def check_arguments():
+    if len(sys.argv) != 2:
+        print('Usage: python3 %s <zephyr.bin>' % sys.argv[0])
```
While you're writing this out, including a one line "what the script is for" help would be good too.

sathishkuttan replied:
I added it in the file below the header.

dbkinder reviewed (approved):
doc changes (and comments added to the Python scripts) LGTM.
